### ISSUE TITLE
"bin" symlinks not created for transitive dependencies when hoisting

### ISSUE THREAD

**Issue by chrisblossom** [lerna/lerna#1130]
## Expected Behavior

Lerna bootstrap should create node_modules/.bin symlinks for nested dependencies that have a bin specified in their package.json. Both ``npm`` and ``yarn`` do this when installing a package normally.

Related issues: https://github.com/lerna/lerna/issues/274 https://github.com/yarnpkg/yarn/issues/2874

## Current Behavior

Lerna bootstrap does not create node_modules/.bin symlinks for nested dependencies that have a bin specified in their package.json.

## Steps to Reproduce (for bugs)

Example: [chrisblossom/lerna-issue-1130](https://github.com/chrisblossom/lerna-issue-1130)

1. Create packges ``one`` and ``two``
2. Add ``two`` as a dependency of ``one``
3. Add ``eslint`` as a dependency of ``two``
4. Check to see if ``one/node_modules/.bin/eslint`` symlink exists.

<details><summary>lerna.json</summary><p>

```json
{
  "lerna": "2.5.1",
  "packages": [
    "packages/*"
  ],
  "version": "1.0.0",
  "command": {
    "bootstrap": {
      "hoist": true
    }
  }
}
```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.5.1 |
| `npm --version`  | 5.5.1 |
| `yarn --version` | 1.3.2 |
| `node --version` | 8.9.1 |

| OS | Version |
| --- | --- |
| macOS Sierra | 10.13.1 |

**Comment by evocateur** [comment]
I'm not a huge fan of relying on implicit dependencies' executables, but I understand it's a legacy thing that some direct dependencies will employ. I don't see this becoming a feature of lerna, chiefly because I'd rather focus on making the npm5 "linklocal" behavior of `file:` dependency specifiers fully-supported, which would then let the package managers manage the packages much better than lerna ever could.

In the hoisting context, those bins are generally present at the root anyway, and `lerna exec` already adds them to the `PATH` when called. That is, `lerna exec -- eslint .` in the provided example would work because `./node_modules/.bin/eslint` exists due to the hoisted bootstrap. This doesn't help external tools that call those transitive dependency executables, of course. That's why I generally recommend condensing all the build/test dependencies in the root, and call them from the root, instead of the leaves.

**Comment by chrisblossom** [comment]
@evocateur thanks for the quick response! I do not believe this is a legacy feature. I agree that library authors are often better off using a ``peerDependency``, but there are legitimate reasons for doing this: https://github.com/yarnpkg/yarn/issues/2874#issuecomment-285182867 and https://github.com/yarnpkg/yarn/issues/2874#issuecomment-294706316 both describe my use case and.

To be clear, are you saying this should be fixed with npm5's "linklocal" behavior? I would be open to starting an npm issue if that would help.

**Comment by evocateur** [comment]
> To be clear, are you saying this should be fixed with npm5's "linklocal" behavior? I would be open to starting an npm issue if that would help.

Actually, this "linklocal" behavior is already in npm5: https://github.com/npm/npm/pull/15900

I was just hoping at some point that lerna could support relative directory links out of the box, and largely fix these sorts of issues (though maybe it's in the same bucket as yarn, in this case? I have no idea). It's a much shorter path, instead of re-implementing even more of npm's internals (which I really, really don't want to do).

To summarize a different way: Lerna's philosophy, whenever possible, is to delegate directly to the underlying package manager for all relevant actions. Until npm extracts a package[1] to make transitive collection of bins for a given dependency tree (relatively) easy, I don't think it makes much sense devoting a ton of time and effort at cobbling together our own (cross-platform?!) solution.

[1]: https://github.com/zkat/cipm has been driving some great extractions, lately, and I'm hopeful that a "package manager toolkit" becomes a thing in the near future.

**Comment by MrCheater** [comment]
+1

**Comment by zxqfox** [comment]
Actually, I didn't pay attention before to this problem but decided now to make an issue and found this one.

In my case I'm using npm 5 and just want to run hoisted binaries in subpackages without changing PATH env.
Like running `ava` from `./packages/one` cwd when `ava` binary exists in `./node_modules/.bin/ava`.

As temporary solution I can add `../../node_modules/.bin` to env PATH but it's risky. And looks like this issue not really about my problem but similar one.
Any ideas?

**Comment by evocateur** [comment]
If you’re calling ava from an npm script in the hoisted subpackage, just add ava to the devDependencies and `lerna bootstrap --hoist` will symlink it appropriately. The fact that a transitive dependency bin is linked in `node_modules/.bin` is not the same as a guarantee that you should be able to access it without specifying an explicit dependency relationship.

**Comment by zxqfox** [comment]
@evocateur Thank you. Feels logical. I'll stick to explicit defining dev deps in each subpackage for now.
